### PR TITLE
Rabbit Streams Additional Tests, Update ConfigureRabbitOptions, Add SpringBootEnv

### LINKED ISSUE #286: Integrate Steeltoe Streams with Spring Cloud Data Flow
We need to create the integration with Spring Cloud Data Flow so Steeltoe streams and SCDF are interoperable.

### LINKED ISSUE #624: Rabbit Binder tests
Some of the tests on the inherited classes need to be ported from:[PartitionCapableBinderTests]( https://github.com/spring-cloud/spring-cloud-stream/blob/40971567d5d7f529caa1c9350ae24968cfa21609/spring-cloud-stream-binder-test/src/main/java/org/springframework/cloud/stream/binder/PartitionCapableBinderTests.java
), & [AbstractBinderTests](https://github.com/spring-cloud/spring-cloud-stream/blob/40971567d5d7f529caa1c9350ae24968cfa21609/spring-cloud-stream-binder-test/src/main/java/org/springframework/cloud/stream/binder/AbstractBinderTests.java)

Some of the tests do not apply and need to be documented (ex. CustomBatchingStrategy and Batching with SMLC is not supported with DMLC(AFAIK).

### PR DESCRIPTION
**Pull request by hananiel**
Additional tests in AbstractBinderTests , PartitionCapableBinderTests (resolves #624)
Update the ConfigureRabbitOptions to update RabbitOptions correctly 
Add SpringBootEnv to parse SPRING_APPLICATION_JSON for SCDF (uses #632)(resolves #286)

### CODE REVIEW

File: src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs [hunk 1]
```diff
@@ -244,17 +246,42 @@ public static IServiceCollection ConfigureRabbitOptions(this IServiceCollection
                 .Bind(config.GetSection(RabbitOptions.PREFIX))
                 .Configure<IServiceProvider>((options, provider) =>
                 {
-                    var connectionFactory = provider.GetService<RC.IConnectionFactory>() as RC.ConnectionFactory;
-
-                    if (connectionFactory is not null)
+                    using (var scope = provider.CreateScope())
                     {
-                        options.Addresses = $"{connectionFactory.HostName}:{connectionFactory.Port}";
+                        var connectionFactory = scope.ServiceProvider.GetService<RC.IConnectionFactory>() as RC.ConnectionFactory;
```
Comment (hananiel) [inline src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs:251]: @kenswan see here.

File: src/Messaging/test/RabbitMQ.Test/Extensions/RabbitServiceExtensionsTest.cs [hunk 2]
```diff
@@ -537,9 +537,9 @@ public void ConfigureRabbitOptions_OverrideAddressWithServiceInfo()
             var provider = services.BuildServiceProvider();
             var rabbitOptions = provider.GetRequiredService<IOptions<RabbitOptions>>().Value;
 
-            Assert.Equal(appsettings[usernamePrefix], rabbitOptions.Username);
-            Assert.Equal(appsettings[passwordPrefix], rabbitOptions.Password);
-            Assert.Equal($"192.168.0.90:3306", rabbitOptions.Addresses);
+            Assert.Equal("Dd6O1BPXUHdrmzbP", rabbitOptions.Username);
```
Comment (hananiel) [inline src/Messaging/test/RabbitMQ.Test/Extensions/RabbitServiceExtensionsTest.cs:540]: @kenswan

File: src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs [hunk 3]
```diff
@@ -244,17 +246,42 @@ public static IServiceCollection ConfigureRabbitOptions(this IServiceCollection
...
+                            options.Username = connectionFactory.UserName;
+                            options.Password = connectionFactory.Password;
+                        }
                     }
                 });
 
             return services;
         }
 
+        public static IServiceCollection ConfigureRabbitOptionsForCloudFoundry(this IServiceCollection services, IConfiguration config)
```
Comment (TimHess) [inline src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs:267]: Is this an alternative to `ConfigureRabbitOptions` or might there be some case where they are used together?

I'm not sure that `ForCloudFoundry` is ideal here as settings in `rabbitmq:client` would also be picked up... What about `ConfigureRabbitOptionsFromConnectors`?

Adding xml comments to all public methods in this class would be a really good idea and help clarify questions like the above.

File: src/Messaging/test/RabbitMQ.Test/Extensions/RabbitServiceExtensionsTest.cs [hunk 4]
```diff
@@ -581,5 +581,7 @@ public void AddRabbitConnectionFactory_AddRabbitConnector()
...
..."vhost\": \"9cd41ebb-ff50-42b7-83f7-646291498fd8\" }, \"management\": { \"host\": \"192.168.8.36\", \"hosts\": [ \"192.168.8.36\" ], \"password\": \"HDv2c37a88vn_gS0Oyw7clwnXtfKxp9w\", \"path\": \"/api/\", \"port\": 15672, \"ssl\": false, \"uri\": \"http://1c6e4609-7685-415a-8d6a-4b11645330f7:HDv2c37a88vn_gS0Oyw7clwnXtfKxp9w@192.168.8.36:15672/api/\", \"uris\": [ \"http://1c6e4609-7685-415a-8d6a-4b11645330f7:HDv2c37a88vn_gS0Oyw7clwnXtfKxp9w@192.168.8.36:15672/api/\" ], \"username\": \"1c6e4609-7685-415a-8d6a-4b11645330f7\" }, \"mqtt\": { \"host\": \"192.168.8.36\", \"hosts\": [ \"192.168.8.36\" ], \"password\": \"HDv2c37a88vn_gS0Oyw7clwnXtfKxp9w\", \"port\": 1883, \"ssl\": false, \"uri\": \"mqtt://9cd41ebb-ff50-42b7-83f7-646291498fd8%3A1c6e4609-7685-415a-8d6a-4b11645330f7:HDv2c37a88vn_gS0Oyw7clwnXtfKxp9w@192.168.8.36:1883\", \"uris\": [ \"mqtt://9cd41ebb-ff50-42b7-83f7-646291498fd8%3A1c6e4609-7685-415a-8d6a-4b11645330f7:HDv2c37a88vn_gS0Oyw7clwnXtfKxp9w@192.168.8.36:1883\" ], \"username\": \"9cd41ebb-ff50-42b7-83f7-646291498fd8:1c6e4609-7685-415a-8d6a-4b11645330f7\" } }, \"username\": \"1c6e4609-7685-415a-8d6a-4b11645330f7\" }, \"syslog_drain_url\": null, \"volume_mounts\": [ ]}]}";
```
Comment (TimHess) [inline src/Messaging/test/RabbitMQ.Test/Extensions/RabbitServiceExtensionsTest.cs:585]: This doesn't appear to be used

File: src/Stream/src/Abstractions/Binder/IBinder.cs [hunk 5]
```diff
@@ -14,7 +14,7 @@ namespace Steeltoe.Stream.Binder
     /// channel adapter, another message channel, etc.
     /// </summary>
     /// TODO:  Figure out Closable/Disposable usage
```
Comment (TimHess) [inline src/Stream/src/Abstractions/Binder/IBinder.cs:16]: Is this TODO still relevant?

File: src/Stream/src/BinderRabbitMQ/Config/RabbitConsumerOptions.cs [hunk 6]
```diff
@@ -54,6 +54,8 @@ public RabbitConsumerOptions()
 
         public string AnonymousGroupPrefix { get; set; }
 
+        public bool? IsEnableBatching { get; set; }
```
Comment (TimHess) [inline src/Stream/src/BinderRabbitMQ/Config/RabbitConsumerOptions.cs:57]: Does it make sense for this to be nullable? Why not just `bool`?

File: src/Stream/src/BinderRabbitMQ/Config/RabbitProducerOptions.cs [hunk 7]
```diff
@@ -43,6 +43,8 @@ public RabbitProducerOptions()
 
         public string ConfirmAckChannel { get; set; }
 
+        public bool? UseConfirmHeader { get; set; }
```
Comment (TimHess) [inline src/Stream/src/BinderRabbitMQ/Config/RabbitProducerOptions.cs:46]: Does it make sense for this to be nullable? Why not just `bool`?

File: src/Stream/src/BinderRabbitMQ/RabbitMessageChannelBinder.cs [hunk 8]
```diff
@@ -137,10 +143,13 @@ protected override IMessageHandler CreateProducerMessageHandler(IProducerDestina
...
             var exchangeName = destination.Name;
             var destinationName = string.IsNullOrEmpty(prefix) ? exchangeName : exchangeName[prefix.Length..];
-            var endpoint = new RabbitOutboundEndpoint(ApplicationContext, BuildRabbitTemplate(extendedProperties, errorChannel != null), _logger)
-            {
-                ExchangeName = exchangeName
-            };
+            var endpoint = new RabbitOutboundEndpoint(
+                ApplicationContext,
+                BuildRabbitTemplate(extendedProperties, errorChannel != null || extendedProperties.UseConfirmHeader.GetValueOrDefault()),
+                _logger)
```
Comment (TimHess) [inline src/Stream/src/BinderRabbitMQ/RabbitMessageChannelBinder.cs:149]: This line is getting complicated, would something like this make it easier to read?
```suggestion
            var template = BuildRabbitTemplate(extendedProperties, errorChannel != null || extendedProperties.UseConfirmHeader.GetValueOrDefault());
            var endpoint = new RabbitOutboundEndpoint(ApplicationContext, template, _logger)
```

File: src/Stream/test/BinderRabbitMQ.Test/RabbitBinderTests.cs [hunk 9]
```diff
@@ -1843,12 +1892,26 @@ public void TestPolledConsumerWithDlqRePub()
             binding.Unbind();
         }
 
-        [Fact]
-        public void TestAnonymousGroup()
+        // TestCustomBatchingStrategy - not supported in Steeltoe
```
Comment (TimHess) [inline src/Stream/test/BinderRabbitMQ.Test/RabbitBinderTests.cs:1895]: Is this comment related to a test that wasn't ported? Without context I don't understand what it's here for

File: src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs (hunk 3 above)
Comment (hananiel) [reply src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs:267]: This needs to be removed. Thanks!

File: src/Stream/src/BinderRabbitMQ/Config/RabbitConsumerOptions.cs (hunk 6 above)
Comment (hananiel) [reply src/Stream/src/BinderRabbitMQ/Config/RabbitConsumerOptions.cs:57]: I am just keeping all the options consistent, in this case, I don't know there is a difference.

File: src/Stream/src/BinderRabbitMQ/Config/RabbitProducerOptions.cs (hunk 7 above)
Comment (hananiel) [reply src/Stream/src/BinderRabbitMQ/Config/RabbitProducerOptions.cs:46]: I am just keeping all the options consistent

File: src/Messaging/src/RabbitMQ/Steeltoe.Messaging.RabbitMQ.csproj [hunk 10]
```diff
@@ -16,6 +16,7 @@
   <ItemGroup>
     <ProjectReference Include="..\..\..\Common\src\Common.Expression\Steeltoe.Common.Expression.csproj" />
     <ProjectReference Include="..\..\..\Common\src\Common.RetryPolly\Steeltoe.Common.RetryPolly.csproj" />
+    <ProjectReference Include="..\..\..\Connectors\src\ConnectorBase\Steeltoe.Connector.ConnectorBase.csproj" />
```
Comment (TimHess) [inline src/Messaging/src/RabbitMQ/Steeltoe.Messaging.RabbitMQ.csproj:19]: I don't see this used anywhere

Comment (hananiel) [reply src/Messaging/src/RabbitMQ/Steeltoe.Messaging.RabbitMQ.csproj:19]: Might have added while troubleshooting - let me remove and see.

Comment (TimHess) [reply src/Messaging/src/RabbitMQ/Steeltoe.Messaging.RabbitMQ.csproj:19]: J/k it's required and used in RabbitServicesExtensions

File: src/Stream/test/BinderRabbitMQ.Test/RabbitBinderTests.cs [hunk 11]
```diff
@@ -1843,12 +1892,26 @@ public void TestPolledConsumerWithDlqRePub()
...
+            var dest = GetExpectedRoutingBaseDestination($"'part{delimiter}0'", "test") + " + '-' + Headers['" + BinderHeaders.PARTITION_HEADER + "']";
+
+            Assert.Contains(dest, routingExpression);
         }
 
+        protected override string GetExpectedRoutingBaseDestination(string name, string group) => name;
+
+        protected override bool UsesExplicitRouting() => true;
+
         private void TestAutoBindDLQPartionedConsumerFirstWithRepublishGuts(bool withRetry)
```
Comment (TimHess) [inline src/Stream/test/BinderRabbitMQ.Test/RabbitBinderTests.cs:1915]: This looks like it's used by 2 tests, one passes true and the other false - should the combination be converted to a theory with inline data instead?

Comment (hananiel) [reply src/Stream/test/BinderRabbitMQ.Test/RabbitBinderTests.cs:1915]: I thought of it, but it seemed best to keep name parity with the tests on the spring cloud stream binder side.

File: src/Configuration/test/PlaceholderBase.Test/PlaceholderResolverConfigurationExtensionsTest.cs [hunk 12]
```diff
@@ -25,6 +25,7 @@ public void AddPlaceholderResolver_ThrowsIfConfigBuilderNull()
         }
 
         [Fact]
+        [Trait("Category", "SkipOnMacOS")]
         public void AddPlaceholderResolver_ThrowsIfConfigNull()
```
Comment (hananiel) [inline src/Configuration/test/PlaceholderBase.Test/PlaceholderResolverConfigurationExtensionsTest.cs:29]: This failed twice in a row on mac net5.0 build. But it wasn't a failure, the test runner crashed with this as most likely failure.  see: https://dev.azure.com/SteeltoeOSS/Steeltoe/_build/results?buildId=5442&view=logs&j=57fb7e2e-87ff-5762-77a2-e1b9cb7a53c7&t=57c60dbc-0a83-516a-6c4f-11e50ea12df3

File: src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs [hunk 13]
```diff
@@ -244,11 +244,18 @@ public static IServiceCollection ConfigureRabbitOptions(this IServiceCollection
...
+
+                        if (connectionFactory is not null)
+                        {
+                            options.Addresses = $"{connectionFactory.UserName}:{connectionFactory.Password}@{connectionFactory.HostName}:{connectionFactory.Port}";
+                            options.VirtualHost = connectionFactory.VirtualHost;
+                            options.Host = connectionFactory.HostName;
+                            options.Username = connectionFactory.UserName;
+                            options.Password = connectionFactory.Password;
+                        }
                     }
```
Comment (hananiel) [inline src/Messaging/src/RabbitMQ/Extensions/RabbitServicesExtensions.cs:259]: It is somewhat accurate - but we have a test that passes in windows , not linux on build server

File: src/Stream/src/StreamBase/Extensions/StreamServicesExtensions.cs [hunk 14]
```diff
@@ -105,5 +110,36 @@ public static void AddStreamServices(this IServiceCollection services, IConfigur
...
+            var loggerFactory = services.BuildServiceProvider().GetService<ILoggerFactory>();
+            var logger = loggerFactory?.CreateLogger("Steeltoe.Stream.Extensions.SafeAddRabbitMQConnection");
+
+            try
+            {
+                services.AddRabbitMQConnection(configuration);
+            }
+            catch (Exception ex)
+            {
+                logger?.LogDebug(ex, ex.Message);
```
Comment (TimHess) [inline src/Stream/src/StreamBase/Extensions/StreamServicesExtensions.cs:141]: Should this be logging as a warning instead of just debug?

Comment (hananiel) [reply src/Stream/src/StreamBase/Extensions/StreamServicesExtensions.cs:141]: Sure yes, I had it originally as error, but wasn't sure if it was somehow tripping the test runner.

File: src/Stream/test/StreamBase.Test/Extensions/HostBuilderExtensionsTest.cs [hunk 15]
```diff
@@ -13,19 +16,13 @@
...
-            var hostBuilder = Host.CreateDefaultBuilder().
-                  AddStreamsServices<SampleSink>();
-            var server = hostBuilder
-                .AddStreamsServices<SampleSink>()
-                .Build();
-
-            var service = server.Services.GetService<IHostedService>();
-            Assert.NotNull(service);
+            var hostBuilder = Host.CreateDefaultBuilder().AddStreamsServices<SampleSink>();
+            Assert.NotNull(hostBuilder);
```
Comment (TimHess) [inline src/Stream/test/StreamBase.Test/Extensions/HostBuilderExtensionsTest.cs:25]: This test should probably assert that `AddStreamsServices<T>` did something (besides not throw an exception)

File: azure-pipelines.yml [hunk 16]
```diff
@@ -33,7 +33,7 @@ jobs:
     inputs:
       command: test
       projects: src/Steeltoe.All.sln
-      arguments: '--blame --no-build -c $(buildConfiguration) -v diag -f netcoreapp3.1 --filter "Category!=SkipOnLinux&Category!=Integration&FullyQualifiedName!~CircuitBreaker"'
+      arguments: '--blame --no-build -c $(buildConfiguration) -v n -f netcoreapp3.1 --filter "Category!=SkipOnLinux&Category!=Integration&FullyQualifiedName!~CircuitBreaker&FullyQualifiedName!~Steeltoe.Stream"'
```
Comment (TimHess) [inline azure-pipelines.yml:36]: Why exclude these from this step? `&FullyQualifiedName!~Steeltoe.Stream`

File: src/Connectors/src/ConnectorCore/RabbitMQProviderServiceCollectionExtensions.cs [hunk 17]
```diff
@@ -10,6 +10,7 @@
 using Steeltoe.Connector.Services;
 using System;
 using System.Linq;
+using System.Reflection;
```
Comment (TimHess) [inline src/Connectors/src/ConnectorCore/RabbitMQProviderServiceCollectionExtensions.cs:13]: If you push more changes to other files before merging, I would back out changes to this file as there's no reason for there to be any changes in here

File: src/Stream/src/StreamBase/Extensions/HostBuilderExtensions.cs [hunk 18]
```diff
@@ -17,31 +19,12 @@ public static class HostBuilderExtensions
...
-                services.AddStreamCoreServices(configuration);
-
-                services.AddBinderServices(configuration);
-                services.AddSourceStreamBinding();
-                services.AddSinkStreamBinding();
-
-                services.AddEnableBinding<T>();
-            };
+                .ConfigureAppConfiguration(cb => cb.AddSpringBootEnv())
+                .ConfigureServices((context, services) =>
```
Comment (TimHess) [inline src/Stream/src/StreamBase/Extensions/HostBuilderExtensions.cs:23]: This looks better than calling ConfigureServices twice 👍🏼

File: azure-pipelines.yml (hunk 16 above)
Comment (hananiel) [reply azure-pipelines.yml:36]: I was reverting to an earlier version - should have just reverted to the `main` version.

File: src/Configuration/test/PlaceholderBase.Test/PlaceholderResolverConfigurationExtensionsTest.cs (hunk 12 above)
Comment (TimHess) [reply src/Configuration/test/PlaceholderBase.Test/PlaceholderResolverConfigurationExtensionsTest.cs:29]: Weird, it looks like the next build after that failed on `AddPlaceholderResolver_WithConfiguration_ReturnsNewConfiguration`. In all cases the stack trace of the exception mentions a `FileSystemWatcher`, so it probably wasn't either of these tests, but rather one of the (4) others in that file that deals with temp files.

**Review by TimHess — APPROVED** [review]
